File: main/domain/src/algorithm/dijkstra.cpp
// Copyright (c) fangguojian. All rights reserved.
// author: fangguojian (destinydesigner@qq.com)
// date: 2014-08-20
// summary: dijkstra.cpp
#include <algorithm>
#include <vector>
#include <map>
#include "comm/define.h"
#include "comm/log.h"
#include "model/town.h"
#include "model/town_list.h"
#include "model/route.h"
#include "model/route_list.h"
#include "algorithm/dijkstra.h"

using assignment::trains::comm::CLog;
using assignment::trains::comm::INFINITE_DISTANCE;
using assignment::trains::domain::CTown;
using assignment::trains::domain::CTownList;
using assignment::trains::domain::CTownIterator;
using assignment::trains::domain::CDistanceFromSrc;
using assignment::trains::domain::CDijkstra;

CDistanceFromSrc::CDistanceFromSrc(long dest, long distance):
    m_dest(dest),
    m_distance(distance)
{
}

long CDistanceFromSrc::getDistance() const
{
    return m_distance;
}

long CDistanceFromSrc::getDest() const
{
    return m_dest;
}

bool CDistanceFromSrc::operator<(const CDistanceFromSrc& node) const
{
    return m_distance > node.getDistance();
}

CDijkstra::CDijkstra(const ITownsMap& map):
    m_pTownMap(&map)
{
}

void CDijkstra::setMap(const ITownsMap& map)
{
    m_pTownMap = &map;
}

long CDijkstra::getMinimum(long src, long dest)
{
    MY_LOG->log("%s. src: %ld, dest: %ld\n", __FUNCTION__, src, dest);
    std::vector<CDistanceFromSrc> nodeQueue;
    std::map<long, long> distanceFromSrcTo;
    std::map<long, bool> visitFlag;

    CTownIterator townIter = m_pTownMap->getAllTowns().getIterator();
    MY_LOG->log("begin town:\n");
    while(townIter.hasNext())
    {
        const ITown& town = townIter.next();
        distanceFromSrcTo[town.id()] = INFINITE_DISTANCE;
        visitFlag[town.id()] = false;
        MY_LOG->log("%ld\n", town.id());
    }
    MY_LOG->log("end town:\n");

    CDistanceFromSrc oDistanceFromSrc(src, 0);
    nodeQueue.push_back(oDistanceFromSrc);
    std::make_heap(nodeQueue.begin(), nodeQueue.end());
    distanceFromSrcTo[src] = 0;

    bool resetSrcFlag = false;
    while(nodeQueue.size() > 0)
    {
        long currentNode = nodeQueue[0].getDest();
        MY_LOG->log("current: %lu\n", currentNode);
        std::pop_heap(nodeQueue.begin(), nodeQueue.end());
        nodeQueue.pop_back();

        MY_LOG->log(
            "current adjacent: %s\n",
            m_pTownMap->getAllAdjacentRoutesBySource(currentNode)
            .ToString().c_str());
        CRouteIterator
            oRouteIter = m_pTownMap->getAllAdjacentRoutesBySource(currentNode)
            .getIterator();
        while(oRouteIter.hasNext())
        {
            CRoute oRoute = oRouteIter.next();
            long nextNode = oRoute.getDestination().id();
            MY_LOG->log("next: %lu\n", nextNode);
            if(distanceFromSrcTo[currentNode] + oRoute.getLength()
               < distanceFromSrcTo[nextNode])
            {
                distanceFromSrcTo[nextNode]
                    = distanceFromSrcTo[currentNode] + oRoute.getLength();
                nodeQueue.push_back(
                    CDistanceFromSrc(nextNode, distanceFromSrcTo[nextNode]));
                std::push_heap(nodeQueue.begin(), nodeQueue.end());
                MY_LOG->log(
                    "after  [%lu, %lu] => %lu\n",
                    currentNode, nextNode, distanceFromSrcTo[nextNode]);
            }
        }

        if (currentNode == src && !resetSrcFlag)
        {
            distanceFromSrcTo[src] = INFINITE_DISTANCE;
            resetSrcFlag = true;
        }
    }

    if (distanceFromSrcTo.find(dest) != distanceFromSrcTo.end()
        && distanceFromSrcTo[dest] != INFINITE_DISTANCE)
    {
        return distanceFromSrcTo[dest];
    }

    return INFINITE_DISTANCE;
}

File: main/domain/include/algorithm/Dijkstra.java
package algorithm;

import java.util.AbstractQueue;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Map.Entry;

import comm.Define;
import domain.TownsMap;

public class Dijkstra {
	private TownsMap map;
	
	public Dijkstra() {
		map = new TownsMap();
	}
	
	public Dijkstra(TownsMap m) {
		setMap(m);
	}
	
	private class MinimumDistance {
		public MinimumDistance() {}
		public MinimumDistance(int dest, int distance) {
			this.dest = dest;
			this.distance = distance;
		}
		private Integer dest;
		private Integer distance;
		public Integer getDest() {
			return dest;
		}
		
		public Integer getDistance() {
			return distance;
		}
	}
	
	private class MinimumDistanceComparator implements Comparator<MinimumDistance> {
		@Override
		public int compare(MinimumDistance lhv, MinimumDistance rhv) {
			if (lhv.getDistance() < rhv.getDistance())
				return -1;
			else if (lhv.getDistance() > rhv.getDistance())
				return 1;
			else
				return 0;
		}
	}
	
	public int calculate(int src, int dest) {
		AbstractQueue<MinimumDistance> nodeQueue = new PriorityQueue<MinimumDistance>(
				new MinimumDistanceComparator());
		Map<Integer, Integer> distanceFromSrcTo = new HashMap<Integer, Integer>();
		Map<Integer, Boolean> visitFlag = new HashMap<Integer, Boolean>();
		for(Integer town : getMap().getAllTowns()) {
			distanceFromSrcTo.put(town, Define.INFINITE_DISTANCE);
			visitFlag.put(town, false);
		}
		nodeQueue.add(new MinimumDistance(src, 0));
		distanceFromSrcTo.put(src, 0);
		//System.out.println("Dijkstra: " + src + ", " + dest);
		
		Boolean resetSrcFlag = false;
		while(nodeQueue.size() > 0) {
			Integer currentNode = nodeQueue.poll().getDest();
			//System.out.println(currentNode);
			Map<Integer, Integer> adjacentNodeOfCurrentNode = getMap().getAdjacentOf(currentNode);
			for(Entry<Integer, Integer> entry : adjacentNodeOfCurrentNode.entrySet()) {
				Integer nextNode = entry.getKey();
				//System.out.println("next: " + nextNode);
				if(distanceFromSrcTo.get(currentNode) + entry.getValue() < distanceFromSrcTo.get(nextNode)) {
					distanceFromSrcTo.put(
							nextNode, distanceFromSrcTo.get(currentNode) + entry.getValue());
					nodeQueue.add(
							new MinimumDistance(nextNode, distanceFromSrcTo.get(nextNode)));
					/*System.out.println(
							"[" + currentNode + ", " + nextNode + "] => " + 
							distanceFromSrcTo.get(nextNode));*/
				}
			}
			if (currentNode == src && !resetSrcFlag) {
				distanceFromSrcTo.put(src, Define.INFINITE_DISTANCE);
				resetSrcFlag = true;
			}
		}
		return distanceFromSrcTo.get(dest);
	}
	private TownsMap getMap() {
		if (this.map == null) {
			this.map = new TownsMap();
		}
		return map;
	}
	public void setMap(TownsMap map) {
		this.map = map;
	}
}
